docs/spec/lightclient/detection/req-ibc-detection.md

```diff
@@ -0,0 +1,296 @@
+   handler for chain A (on chain B)
+   is on a different branch than the relayer
+
+
+- in both detection scenarios, the relayer should submit evidence to 
+  full nodes of chain A where there is a fork. As we assume a fullnode
+  has a complete list of blocks, it is sufficient to send "Bucky's
+  evidence" (https://github.com/tendermint/tendermint/issues/5083),
+  that is, 
+     - two lightblocks from different branches + 
+	 - a lightblock (perhaps just a height) from which both blocks 
+	   can be verified.
+  
+- in the scenario 2., the relayer must feed the A-handler (on chain B)
+  a proof of a fork on A so that chain B can react accordingly
+  
+### Handler requirement
+  
+- there are potentially many relayers, some correct some faulty
+
+- a handler cannot trust the information provided by the relayer, 
+  but must verify 
+  (Доверя́й, но проверя́й)
+
+- in case of a fork, we accept that the handler temporarily stores
+  headers (tagged as verified). 
+  
+- eventually, a handler should be informed
+ (`checkMisbehaviourAndUpdateState`)
+ by some relayer that it has
+  verified a header from a fork. Then the handler should do what is
+ required by IBC in this case (stop?)
+
+### Challenges in the handler requirement
+
+- handlers and relayers work on different lightstores. In principle
+  the lightstore need not intersect in any heights a priori
+
+- if a relayer  sees a header *h* it doesn't know at a handler (`queryChainConsensusState`), the
+  relayer needs to
+  verify that header. If it cannot do it locally based on downloaded
+  and verified (trusted?) light blocks, it might need to use
+  `VerifyToTarget` (bisection). To call `VerifyToTarget` we might keep
+  *h* in the lightstore. If verification fails, we need to download the
+  "alternative" header of height *h.Height* to generate evidence for
+  the handler.
+  
+- we have to specify what precisely `queryChainConsensusState`
+  returns. It cannot be the complete lightstore. Is the last header enough?
+
+- we would like to assume that every now and then (smaller than the
+  trusting period) a correct relayer checks whether the handler is on a
+  different branch than the relayer. 
+  And we would like that this is enough to achieve
+  the Handler requirement.
+  
+   - here the correctness argument would be easy if a correct relayer is
+     based on a light client with a *trusted* state, that is, a light
+     client who never changes its opinion about trusted. Then if such a
+     correct relayer checks-in with a handler, it will detect a fork, and
+     act in time.
+
+   - if the light client does not provide this interface, in the case of
+     a fork, we need some assumption about a correct relayer being on a
+     different branch than the handler, and we need such a relayer to 
+	 check-in not too late. Also
+     what happens if the relayer's light client is forced to roll-back
+     its lightstore?
+     Does it have to re-check all handlers?
```

Comment:
Perhaps we should maintain a local copy of all headers from handlers, as if handlers were another witness in the light client system?

We'd basically want to know at all times that our light client conforms with those of the handers we know about. If at any point our light client has to roll-back, we should assume that means there's a fork relevant to all those handlers as well.

Basically we want to roughly ensure that our light client and those of all the handlers are basically in sync and in agreement so it's easy to detect forks and make all parties aware of them.

Reply:
> Basically we want to roughly ensure that our light client and those of all the handlers are basically in sync and in agreement so it's easy to detect forks and make all parties aware of them.

This makes a lot of sense. Do you think performance-wise it is reasonable to maintain such a tight synchronization?

Reply:
> This makes a lot of sense. Do you think performance-wise it is reasonable to maintain such a tight synchronization?

Not sure. Syncing the on-chain clients to every header a relayer has is probably excessive, but we might want relayers to download and verify every header that gets updated on-chain. I guess whenever the relayer receives an UpdateClient event from a chain, it should tell its own light client to fetch and verify that height so it can be on the lookout for forks. But the relayer should always make sure its at least as up-to-date as the onchain client

Reply:
This is exactly what I had in mind. Whenever relayer sees UpdateClient event, it fetches and verifies on chain data and input this info to the light client so it can verify it.